coderd/unhanger/detector.go

```diff
@@ -0,0 +1,332 @@
+			// our messages are in the latest stage.
+			logs, err := db.GetProvisionerLogsAfterID(ctx, database.GetProvisionerLogsAfterIDParams{
+				JobID:        job.ID,
+				CreatedAfter: 0,
+			})
+			if err != nil {
+				log.Warn(ctx, "get logs for hung job", slog.Error(err))
+				continue
+			}
+			logStage := ""
+			if len(logs) != 0 {
+				logStage = logs[len(logs)-1].Stage
+			}
+			if logStage == "" {
+				logStage = "Unknown"
+			}
+
+			// Insert the messages into the build log.
+			insertParams := database.InsertProvisionerJobLogsParams{
+				JobID: job.ID,
+			}
+			now := database.Now()
+			for i, msg := range HungJobLogMessages {
+				// Set the created at in a way that ensures each message has
+				// a unique timestamp so they will be sorted correctly.
+				insertParams.CreatedAt = append(insertParams.CreatedAt, now.Add(time.Millisecond*time.Duration(i)))
+				insertParams.Level = append(insertParams.Level, database.LogLevelError)
+				insertParams.Stage = append(insertParams.Stage, logStage)
+				insertParams.Source = append(insertParams.Source, database.LogSourceProvisionerDaemon)
+				insertParams.Output = append(insertParams.Output, msg)
+			}
+			newLogs, err := db.InsertProvisionerJobLogs(ctx, insertParams)
+			if err != nil {
+				log.Warn(ctx, "insert logs for hung job", slog.Error(err))
+				continue
+			}
+
+			// Publish the new log notification to pubsub. Use the lowest
+			// log ID inserted so the log stream will fetch everything after
+			// that point.
+			lowestID := newLogs[0].ID
+			data, err := json.Marshal(provisionersdk.ProvisionerJobLogsNotifyMessage{
+				CreatedAfter: lowestID - 1,
+			})
+			if err != nil {
+				log.Warn(ctx, "marshal log notification", slog.Error(err))
+				continue
+			}
+			err = d.pubsub.Publish(provisionersdk.ProvisionerJobLogsNotifyChannel(job.ID), data)
+			if err != nil {
+				log.Warn(ctx, "publish log notification", slog.Error(err))
+				continue
+			}
+
+			// Mark the job as failed.
+			now = database.Now()
+			err = db.UpdateProvisionerJobWithCompleteByID(ctx, database.UpdateProvisionerJobWithCompleteByIDParams{
+				ID:        job.ID,
+				UpdatedAt: now,
+				CompletedAt: sql.NullTime{
+					Time:  now,
+					Valid: true,
+				},
+				Error: sql.NullString{
+					String: "Coder: Build has been detected as hung for 5 minutes and has been terminated.",
+					Valid:  true,
+				},
+				ErrorCode: sql.NullString{
+					Valid: false,
+				},
+			})
+			if err != nil {
+				log.Warn(ctx, "mark job as failed", slog.Error(err))
+				continue
+			}
+
+			// If the provisioner job is a workspace build, copy the
+			// provisioner state from the previous build to this workspace
+			// build.
+			if job.Type == database.ProvisionerJobTypeWorkspaceBuild {
+				build, err := db.GetWorkspaceBuildByJobID(ctx, job.ID)
+				if err != nil {
+					log.Warn(ctx, "get workspace build for workspace build job by job id", slog.Error(err))
+					continue
```

Comment:
This `continue` (and below) also feels a bit suspicious, I don't think we should accept a _maybe_ state where some terminations update the build and others don't due to an error.